6-infrastructure-configuration-management/6.1.3-terraform-modules.md

```diff
@@ -4,6 +4,192 @@ The previous sections should have illustrated how flexible and powerful Terrafor
+  type        = string
+  default     = "10.0.0.0/16"
+}
+
+variable "vpc_azs" {
+  description = "Availability zones for VPC"
+  type        = list(string)
+  default     = ["us-east-2a", "us-east-2b", "us-east-2c"]
+}
+
+variable "vpc_private_subnets" {
+  description = "Private subnets for VPC"
+  type        = list(string)
+  default     = ["10.0.1.0/24", "10.0.2.0/24"]
+}
+
+variable "vpc_public_subnets" {
+  description = "Public subnets for VPC"
+  type        = list(string)
+  default     = ["10.0.101.0/24", "10.0.102.0/24"]
+}
+
+variable "vpc_enable_nat_gateway" {
+  description = "Enable NAT gateway for VPC"
+  type        = bool
+  default     = true
+}
+
+variable "vpc_tags" {
+  description = "Tags to apply to resources created by VPC module"
+  type        = map(string)
+  default = {
+    Terraform   = "true"
+    Environment = "sandbox"
+  }
+}
+```
+
+2. Review the configuration below first. The `  source  = "terraform-aws-modules/vpc/aws"
+` and `  source  = "terraform-aws-modules/ec2-instance/aws"
+` declare the registry and module thats being used for the modules in this configuration.
+
+
+3. Create a new `.tf` file or add the following configuration to your `main.tf` file from previous sections of the boot camp.  
+
+4. For this exercise the `ami` id is hardcoded. A new ID can be found in the AWS console or the AWS CLI. This AWS user guide walks through that process - [Find a Linux AMI](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/finding-an-ami.html#finding-an-ami-console)
+
+?> As mentioned previously, the configuration below creates a new VPC for the EC2 instance to reside in. If you don't want to create a new VPC, you can omit the VPC module from your configuration. For this exercise, the ec2 instance relies on outputs from the VPC module. If you omit the VPC module, `vpc_security_group_ids` and `subnet_id` will have to be hardcoded based on resources available in the AWS account you're deploying into. 
+
+```hcl
+module "vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "3.18.1"
+
+  name = var.vpc_name
+  cidr = var.vpc_cidr
+
+  azs             = var.vpc_azs
+  private_subnets = var.vpc_private_subnets
+  public_subnets  = var.vpc_public_subnets
+
+  enable_nat_gateway = var.vpc_enable_nat_gateway
+
+  tags = var.vpc_tags
+}
+
+module "ec2_instance" {
+  source  = "terraform-aws-modules/ec2-instance/aws"
+  version = "4.3.0"
+
+  count = 1
+  name  = "my-ec2-cluster-${count.index}"
+
+  ami                    = "ami-0c5204531f799e0c6"
+  instance_type          = "t2.micro"
+  vpc_security_group_ids = [module.vpc.default_security_group_id]
+  subnet_id              = module.vpc.private_subnets[0]
+
+  tags = {
+    Terraform   = "true"
+    Environment = "sandbox"
+  }
+}
+```
+
+3. Commit and push these changes to your repository. Push this new configuration to a branch that is configured to run a `terraform plan.` View the results of the plan before proceeding to deploy. 
```

Comment:
Minor question - are there already previous steps somewhere in the modules to ensure that the user has a branch that's configured to run a terraform plan? Or instructions on how to do so? I assume so - but having not gone through these modules myself, just wanted to make sure this wouldn't be a sticking point for anyone :) So this might be an unnecessary question :)

Reply:
Yes in the previous steps we talk about configuring a pipeline to run your terraform. 👍